frontend/src/components/login/LoginForm.tsx

```diff
@@ -0,0 +1,51 @@
+import React, { FormEvent, useState, useContext } from "react";
+import { LoginContext } from "../login/LoginManager";
+
+export const LoginForm: React.FunctionComponent = () => {
+  const [username, setUsername] = useState("");
+  const [password, setPassword] = useState("");
+  const [error, setError] = useState<string>();
+  const loginContext = useContext(LoginContext);
+
+  function tryLogin(event: FormEvent) {
+    event.preventDefault();
+    loginContext.logIn(username, password).then((didLogin) => {
+      if (didLogin) {
+        setError(undefined);
+      } else {
+        setError("Login failed.");
```

Comment:
If you made the `logIn` function throw an exception with a specific error message, you could `setError` inside a `.catch` on that method and display a more helpful message to the user!